### DISCUSSION TITLE
Supabase Middleware and Next js Internationalization with @supabase/ssr

### DISCUSSION THREAD

**Discussion by bennik88** [supabase/supabase#27235] [Questions]
**Problem:**
I'm using Next js Internationalization with the App Router to have multiple languages. Following the [Next js docs](https://nextjs.org/docs/app/building-your-application/routing/internationalization) this is done in the `middleware.tsx `file.
For auth I switched to `@supabase/ssr` which is also setup in the `middleware.tsx`file. The problem is, I'm not sure if I'm following the ideal way to handle the changing route patterns `domain/(locale)/path`as the matcher in the middleware.tsx file excludes path from middleware which seems dangerous to do as the middleware is updating the session and therefore proofing that the user is logged in. But my supabase `/auth` and generals `/api` routes are not in my `/app/[lang]` folder they are in `/app/api` and `/app/auth/confirm` so I have to include them into the mater to work or my fetch requests try to do something like `domain/en/api/pathToMyApi` and for sure I don't want to have the "en" part included. So I added an own exclude matcher to middleware, but it looks like a bad way to solve it and it's not very clean. 

**Question:**
What would be the preferred way to handle this very common scenario? 

**My code:**

```js
// Next JS
import type { NextRequest } from "next/server";
import { NextResponse } from "next/server";
import { headers } from "next/headers";

// Language Routing
import { match } from "@formatjs/intl-localematcher";
import Negotiator from "negotiator";

// Types
import { AvailableLanguagesLowercase } from "@/data/constants";
import { updateSession } from "./utils/supabase/middleware";

let locales = ["en"]; // , "ja" AvailableLanguagesLowercase "ja" // do i have to set "en-US"??
let defaultLocale = "en";

const LOGS = false;
const ENGLISH_ONLY = true;

// Get the preferred locale
function getLocale(request: NextRequest) {
  // English only
  if (ENGLISH_ONLY) return "en";

  // Get headers
  // const headersList = headers()
  // const acceptLanguage = headersList.get("accept-language")
  const headersLanguage = request.headers.get("accept-language");
  if (typeof headersLanguage !== "string") return undefined;

  // Sample headers: let headers = { "accept-language": "en-US,en;q=0.5" };
  let headers = { "accept-language": headersLanguage };
  let languages = new Negotiator({ headers }).languages();

  // Get closest matching language
  const locale = match(languages, locales, defaultLocale);
  return locale;
}

function redirectToLocal(request: NextRequest) {
  // Check if there is any supported locale in the pathname
  const pathname = request.nextUrl.pathname;
  // console.log("pathname: ", pathname);
  const pathnameIsMissingLocale = locales.every(
    (locale) => !pathname.startsWith(`/${locale}/`) && pathname !== `/${locale}`
  );

  // Redirect if there is no locale
  if (pathnameIsMissingLocale) {
    // Get closest matching language
    const locale = getLocale(request);
    if (!locale) return;
    // e.g. incoming request is /products
    // The new URL is now /en-US/products
    return new URL(`/${locale}/${pathname}`, request.url);

    // return NextResponse.redirect(
    //   new URL(`/${locale}/${pathname}`, request.url)
    // );
  }
}

export async function middleware(request: NextRequest) {
  // Define the paths to exclude from language redirects
  const excludedPaths = ["/auth/confirm"];
  // const res = NextResponse.next();

  // Extract the pathname from the request URL
  const { pathname } = request.nextUrl;

  // Check if the pathname is in the excluded paths array
  if (excludedPaths.includes(pathname)) {
    // If the pathname is in the excluded paths, continue without modifying the URL
    if (LOGS) {
      console.log("excluded path");
      console.log("update Session");
    }
    return await updateSession(request);
  }

  // Redirect to locale language
  const url = redirectToLocal(request);
  if (!url) {
    if (LOGS) {
      console.log("no redirect to local ");
      console.log("update Session");
    }
    return await updateSession(request);
  }
  updateSession(request);
  // if (LOGS) console.log("redirect: ", url);
  return NextResponse.redirect(url);
}

// For language matching
export const config = {
  matcher: [
    // Skip all internal paths (_next)
    // Skip all API routes
    // Skip supabase callback route
    // Matcher ignoring `/_next/` and `/api/` and auth // my old method
    // "/((?!api|auth/callback|_next|_next/static|_next/image|favicon.ico).*)",

    // Updated matcher 24.06.06
    /*
     * Match all request paths except for the ones starting with:
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     * Feel free to modify this pattern to include more paths.
     */
    "/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)",
  ],
};
```

**Additional Problem with @supabase/ssr**
Another thing that is super confusing about the superbase docs is that the way of how to handle auth is changing all the time and the documentation and supabase tutorial videos seem to mix these ways of handling auth. So, I honestly don't know which is the source of truth. The newest docs say I have to use `client.ts`, `middleware.ts`, `server.ts`, and `app/auth/confirm` for `@supabase/ssr` but the next js supabase starter who also uses `@supabase/ssr` instead uses `app/auth/callback` and the content of the callback route and confirm route are completely different logic to handle auth. So now I'm worried about having introduced some potential security vulnerabilities because I'm not sure which is the right way to handle it. I have both files now, but my users are now getting automatically logged out after a few hours instead of staying logged in like before.

```
"supabase": "^1.172.2"
"@supabase/auth-ui-react": "^0.4.2",
"@supabase/auth-ui-shared": "^0.1.6",
"@supabase/ssr": "^0.3.0",
"@supabase/supabase-js": "^2.43.4",
"react": "^18.3.1",
"@formatjs/intl-localematcher": "^0.4.0",
```

**Comment by dimitarbarev** [comment]
I would separate this into two concerns rather than trying to encode both of them into the matcher:

1. Which requests need Supabase session refresh
2. Which requests need locale routing

`/api/* and /auth/*` should remain non-localized routes. They should not become /en/api/* or /en/auth/*.

But that doesn't mean they necessarily have to be excluded from the middleware/proxy entirely. The matcher can stay broad enough for Supabase session handling, while the locale redirect is conditionally skipped for those routes.

For example:

```
const NON_LOCALIZED_PREFIXES = ["/api", "/auth"]

function shouldLocalize(pathname: string) {
  return !NON_LOCALIZED_PREFIXES.some(
    prefix => pathname === prefix || pathname.startsWith(`${prefix}/`)
  )
}

export async function middleware(request: NextRequest) {
  const { pathname } = request.nextUrl

  // First let Supabase validate/refresh the session.
  const response = await updateSession(request)

  // API and auth endpoints are intentionally not localized.
  if (!shouldLocalize(pathname)) {
    return response
  }

  const redirectUrl = redirectToLocale(request)

  if (redirectUrl) {
    return NextResponse.redirect(redirectUrl)
  }

  return response
}

export const config = {
  matcher: [
    "/((?!_next/static|_next/image|favicon.ico|.*\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)",
  ],
}
```

I think this is cleaner than putting /api and /auth into the matcher exclusions, because the matcher answers "should middleware execute?", while your locale logic answers "should this URL be localized?". Those aren't the same question.

There is also a bug/race in the original code here:

```
updateSession(request);
return NextResponse.redirect(url);
```

You're calling updateSession() but throwing away its returned response (and you're not awaiting it). If Supabase refreshes the session during that request, any Set-Cookie written to that response can be lost when you return a new NextResponse.redirect().

So when combining locale redirects with Supabase SSR, make sure any cookies produced by updateSession() are also copied to the redirect response, e.g. conceptually:

```
const supabaseResponse = await updateSession(request)

if (redirectUrl) {
  const redirectResponse = NextResponse.redirect(redirectUrl)

  supabaseResponse.cookies.getAll().forEach(cookie => {
    redirectResponse.cookies.set(cookie)
  })

  return redirectResponse
}

return supabaseResponse
```

That could also explain seemingly random logouts after the access token expires: normal requests work until Supabase needs to rotate/refresh the session, then the refreshed cookies are written to a response that never reaches the browser.

Regarding auth/confirm vs auth/callback: they represent different auth flows, so I wouldn't keep both just because different examples contain them.

For the current SSR email-confirmation flow, Supabase documents /auth/confirm with a token_hash, followed by verifyOtp(). A callback using exchangeCodeForSession(code) is used for PKCE/code-based flows such as OAuth. Which endpoint you need depends on which auth flow you're actually initiating.

So my structure would be:

```
app/
  [lang]/
    ...
  api/
    ...
  auth/
    confirm/
      route.ts
    callback/
      route.ts   // only if a flow actually needs it

```
with /api/* and /auth/* deliberately outside [lang].

One more security point: middleware refreshing the session and middleware authorizing access aren't quite the same thing. For authorization, don't rely solely on the presence of the session cookie or on getSession() server-side. Current Supabase guidance recommends validating the token (getClaims(), or getUser() when you need the current user record).

So in short: keep the matcher broad, skip locale rewriting for /api and /auth, preserve Supabase's response cookies when creating a redirect response, and keep authentication/authorization separate from i18n routing.
